### PR TITLE
Telemetry Library for OTG

### PR DESCRIPTION
**Pull request by robshakir**
```
commit 2ae265140dbe3a7482bb0c95d4f284dce95eddc4
Author: Rob Shakir <robjs@google.com>
Date:   Sat Aug 20 16:19:36 2022 -0700

    Add initial telemetry server implementation.
    
     * (A) lwotgtelem/*
       - Add a gNMI server using lemming's gnmit that can host tasks
         that populate OTG telemetry.
     * (A) lwotg/*
       - Add initial file for lwotg which solely defines a Hint type.
```

This PR adds an initial telemetry implementation for magna, which is based on
the lemming gnmit library. Tasks are registered into the gNMI server that
populate telemetry values. Values can also be written to the telemetry server
using a hint channel which is added during this PR.

The initial use for a hint channel is to communicate details of the
configuration of the OTG server to the telemetry daemon for mapping.

### CODE REVIEW

File: lwotgtelem/lwotgtelem.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+	if _, gok := s.hints[group]; !gok {
+		return "", false
+	}
+
+	value, ok = s.hints[group][key]
+	return
+}
+
+// AddTask adds the task t to the current tasks run by the gNMI server. It can be used to
```
Comment (akrentsel) [inline lwotgtelem/lwotgtelem.go:97]: what do the tasks represent?

File: lwotgtelem/lwotgtelem_test.go [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+const (
+	certFile = "testdata/cert.pem"
+	keyFile  = "testdata/key.pem"
+)
+
+func TestNew(t *testing.T) {
+	tests := []struct {
+		desc        string
+		inHostname  string
+		inPaths     []*gpb.Path
```
Comment (akrentsel) [inline lwotgtelem/lwotgtelem_test.go:29]: what is the "in" short for?

**Review by akrentsel — APPROVED** [review]

File: lwotg/lwotg.go [hunk 3]
```diff
@@ -0,0 +1,16 @@
+// Package lwotg implements a lightweight, extensible OTG implementation.
+package lwotg
+
+// Hint is <group, key, value> tuple that can be handed to modules of the
+// OTG implementation to perform their functions. For example, it may be
+// used to communicate mappings between system interfaces and the names that
+// are used in OTG for them.
+type Hint struct {
+	// Group is a string used to specify a name for a set of hints that
+	// are associated one one another.
```
[the remark this answers is not in the document]
Comment (robshakir) [reply lwotg/lwotg.go:10]: Fixed, thanks.

File: lwotg/lwotg.go [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+	Group string
```
[the remark this answers is not in the document]
Comment (robshakir) [reply lwotg/lwotg.go:11]: Good question. At the moment it wasn't super clear what hints would need to be passed between different parts of the server, and with the pattern of having new config and flow handlers be able to be added by users without needing to change the core implementation it seemed like it was nicer to just have something that was flexible here. If there turn out to be categories that make sense in the future, an enum would be a good idea.

File: lwotg/lwotg.go [hunk 5]
```diff
@@ -0,0 +1,16 @@
+// Package lwotg implements a lightweight, extensible OTG implementation.
```
[the remark this answers is not in the document]
Comment (robshakir) [reply lwotg/lwotg.go:1]: Done :-)

File: lwotgtelem/lwotgtelem.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+)
+
+// Server contains the implementation of the gNMI server.
+type Server struct {
+	// c is the base gnmit Collector implementation that is used to store updates.
+	c *gnmit.Collector
+	// GNMIServer is the gNMI gRPC server instance that can be exposed externally.
+	GNMIServer *gnmit.GNMIServer
+
+	// hintCh is a channel that is used to write hints into the telemetry server
```
[the remark this answers is not in the document]
Comment (robshakir) [reply lwotgtelem/lwotgtelem.go:21]: The hints can be written by any part of the system that finds that it has some information that is needed to map telemetry information. They might be handlers of the configuration or flows within the system.

File: lwotgtelem/lwotgtelem.go (hunk 1 above)
Comment (robshakir) [reply lwotgtelem/lwotgtelem.go:97]: Clarified - tasks are just new functions that write telemetry information, they can populate any leaf, or set of leaves.

File: lwotgtelem/lwotgtelem_test.go (hunk 2 above)
Comment (robshakir) [reply lwotgtelem/lwotgtelem_test.go:29]: It's just the convention for data that is input to the tests `inFoo` and `wantBar` tends to be the idiomatic test naming convention.

**Comment by robshakir** [comment]
Thanks for the review @akrentsel!

**Pull request merged**